### ISSUE TITLE
Discussion: Inconsistent function name style

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#582]
This came up in issue #581, and I think merits it's own discussion. From [here](https://github.com/dankelley/oce/issues/581#issuecomment-73111533):

> Doing grep in NAMESPACE indicates that the majority of functions are using camel case, and that argues that the new function should be named oceGrid, not oce.grid. I wish R didn't have these choices in function names. One solution might be to just provide non-camel names for every oce function, in addition to the camel-case ones (using aliases).

> PS. I used to hate the underline notation but now that Hadley is promoting it, I no longer feel that way. But it seems too late to alias hundreds of function names.

To illustrate, a list of all the `oce*`-prefix functions is:
```
       oceApprox,
       oce.as.raw,
       oce.as.POSIXlt,
       oce.axis.POSIXct,
       oceColorsGebco,
       oceColorsPalette,
       oceColorsJet,
       oceColorsTwo,
       oceColors9A,
       oceColors9B,
       oceDebug,
       oceEdit,
       oceFilter,
       oceMagic,
       oceConvolve,
       oceContour,
       oce.plot.ts,
       ocePmatch,
       oceSmooth,
       oceSpectrum,
       oce.write.table,
```

I had always thought of the `oce.*()` functions as being alternate versions of existing ones, like `plot.ts()` and `axis.POSIXct()`, and so the "." allows for consistent naming. In such situations, I think camelCase, e.g. `ocePlot.ts()` is ugly and inconsistent.

On the other hand, the camelCase functions represent functions with no base R equivalent, and so are named as all other oce functions.

There are clearly some inconsistencies with this scheme, as can be seen with things like: `oceApprox()`, `oceFilter()`, `oceConvolve()`, `oceContour()`, `ocePmatch()`, `oceSmooth()`, `oceSpectrum()`, which all have base R equivalents.

An alternative is to use underscore notation everywhere, e.g.:
```
oce_approx
oce_colors_jet
```
which is a bit of a matter of preference (Hadley's views aside) but the issue of oce versions of existing functions that already have dots in the name remains.

It would be a lot of work to make aliases for both camelCase and underscore case (after deciding on the "dots in existing function names" issue), but not impossible.

**Comment by dankelley** [comment]
In view of Hadley's views, I have no doubt that underscore is how this would be coded, if oce were being started today.  Although it would be a lot of work to change things, I really do think the underscore notation is what oce should have in version 1.x, and the 0 at the start of the present version suggests that change could be okay.

Having said that, I believe in supporting existing code. There is a solution, though, and that is that the ``.Rd`` files each get an underscore-type ``alias`` command (which will take a half hour, tops), that ``NAMESPACE`` gets a bunch of new lines (again, maybe a half hour) and that the various ``.R`` functions get followed by lines like ``oce_approx <- oceApprox`` (maybe a few hours, since there are a lot of files and it's not wise to try global change/replace search patterns for such work). Then there would have to be a phase in which the examples in the help pages are switched to the underscore names; that will take probably a solid morning.

This is an estimate of maybe a full day.  With the usual snags, say maybe two days.  In other words, it's a weekend project. The good thing is that this can be done in pieces, and the ``NAMESPACE`` file provides a natural checklist. (What I mean is that this is not the kind of work that requires sustained concentration over a full day, as some bugs do.)

From the above, I conclude that a switch is a good idea. It would make me *really* happy to have the main oce system be underscore based.  Now that Hadley has championed this, it seems to be the future. I think many new packages are going to employ this system, and I think we have a chance for ``oce`` to catch the wave.

This is all the more true since oce is at version 0.x and since whatever gets used in OAR is going to become the standard.

I am not talking about changing argument names, just function names.  It's hard to retain backwards compatibility across argument names. Changing argument names will break code, and since R itself has inconsistent argument names, I don't see it as a tragedy for oce to have inconsistent ones. Still, I'd make new argument names use underscore whenever it made sense.

All of the above is off the top of my head, a sort of one-man conversation. I started off thinking switching to underscore was an okay idea, and now I think it's a very good one.

**Comment by richardsc** [comment]
Some of my thoughts:

1. I generally prefer camelCase. Obviously this is quite personal, so isn't really a valid argument except that I'm official one of the package authors now (cue maniacal laughter). I like that it's more compact, and it doesn't save any keystrokes over underscore notation.

2. Part of my preference for camelCase is that as an Emacs (and ESS) user, the "\_" key is bound by default as a shortcut to " <- " (note the spaces around the assign operator, so that it saves 5 keystrokes). To get an underscore, I have to hit "\_" twice. Not the end of the world, but a shift. I will note that RStudio has it's own different shortcut for assign, so this isn't as much an issue for users who are more editor-challenged. (What's the shortcut in vim? Probably something like ESC-:-a-/-\_ ...)

3. It could be confusing to maintain two systems, especially when (if?) the package grows in popularity once OAR is published. There is a fair bit of legacy code around (e.g. the closed issues on this site), which might be confusing for new users.

4. **Arguments names are all camelCase**, and you've already said that you won't change them. For me, this is the biggest strike against an all out switch to underscore notation. Yes, R has inconsistent argument names, but this would **create a complete style dichotomy between function names and argument names, which for consistency and backward compatibility will have to be maintained forever.** I don't see the value in switching to underscores for functions when the arguments can't be changed. Sure, the package version is still < 1, but it's been on GH since at least 2010 (almost 5 years!), and in pretty widespread use even longer (I don't really know the numbers of users, though).

5. The scientific literature says: "[... camel case style leads to better all around performance ...](https://whathecode.wordpress.com/2011/02/10/camelcase-vs-underscores-scientific-showdown/). Who's going to argue with science?

**Comment by richardsc** [comment]
Another Emacs thing that I thought of (though I know you don't care): hitting M-delete does a backward-delete-word, but hangs on underscores as "word" separators. Which means that to delete `oce_colors_jet` I have to hit M-delete 3 times instead of once for `oceColorsJet`.

**Comment by dankelley** [comment]
First, let me say that this is a splendid example of how open-source development can work. I am labelling this "development" and "archive", so that, even after it's closed, it will be easy to find.

Clark, you've convinced me, particularly in view of the fact that arguments are camel case, that a mass shift to underscores is a bad idea. However, I really like the idea of improving consistency, so my thought at the moment is to take the "odd ones" and create dot equivalents (so that old code works).

**Suggestion (comments please)** I can write up the help pages in such a way that it is clear that the old names are just for compatibility, and that the new names are preferred. I can also say that the old names will go away at a certain date (maybe Jan 2016), unless a certain option is set.  (That way, people have a year to say ``options(oceOldNames=TRUE)`` or whatever ... and a warning would be given for say 2 years.)

If you like the above, maybe you could have a look at the following checklist, which I will tick off when I do things. Since you can edit, please add/subtract function names (in this comment, so things don't get confusing). The list is just from your comment above. 
* [ ] ``oceApprox()``
* [ ] ``oceFilter()``
* [ ] ``oceConvolve()``
* [ ] ``oceContour()``
* [ ] ``ocePmatch()``
* [ ] ``oceSmooth()`` 
* [ ] ``oceSpectrum()`

**Comment by richardsc** [comment]
Good to know that I can be so convincing! Did I also tell you my idea of making me the author of your book once it's written? :)

It occurred to me that a note about style could go in the vignette, highlighting the functions which are *not* camelCase, and indicating the rationale behind that choice.

I like the idea of using dots only for functions for which oce creates an *alternate* version, such as `oce.plot.ts()` and `oce.approx()`, etc.

One distinction to make on whether to use the "standard" camelCase vs dotted function names is whether the function is actually an *alternate* to an existing base R function, or if it simply *has the same name*. In the case of `oceEdit()` vs base R `edit()`, the functions actually do completely different things, so you could make the argument that `oceEdit()` is more appropriate than `oce.edit()`. But perhaps that's all a bit semantic ...

**Comment by dankelley** [comment]
I am still struggling with this name issue. To clear my thoughts, I'll review... with numbers in case Clark wants to comment.

1. I find having two schemes confusing. Users should not have to think that hard before guessing a name.  (I'm exaggerating. If they are using some tools, they can type ``oce`` ``TAB`` ``TAB`` to see the functions, but let's suppose they don't know that.)

2. I would like to retain backwards compatibility, which means aliasing. However, I'd like to state a *preferred* form in the docs, basically because I think that will be easier for users to understand.

3. Counting names of existing functions, and preferring for least disruption to users, the preferred form would be camel case.

4. A negative on camel case is that it is *changing* the names of existing functions; ``Approx`` is not the same as ``approx``. And that argues to have the preferred form be ``oce.approx``, not ``oceApprox``. This could confuse users.

5. I think it comes down to weighing items 3 and 4. Neither choice will affect existing code. Neither choice involves much development time.

**DK vote** on preference:
* [ ] use camelCase as the "preferred" notation
* [x] use dot-name as the "preferred" notation

**CR vote** on preference:
* [ ] use camelCase as the "preferred" notation
* [ ] use dot-name as the "preferred" notation

**Comment by richardsc** [comment]
I voted, but I struggle similarly to you. What we're trying to decide on here is which *inconsistent* scheme for function naming is either

a) *least* inconsistent, or

b) the most obvious for users

It's telling from your grep list that the `oce.*()` scheme came about because of adapting base R functions **that already had dots in the names** (with the exception of `oce.grid()`, which was only just born), a symptom of the fact that even base R is horribly inconsistent in style. For the record, if we chose camelCase, I really don't like the way functions like `oceAs.raw()` and `ocePlot.ts()` looks, **at all**. It's so mixed up in style to be ridiculous.

Anyway, my vote stands because:

1. It is a scheme that distinguishes base R and oce versions of functions

2. shouldn't involve much coding and change.

**Comment by dankelley** [comment]
Good. We voted the same way, so it's easy. I'll make a checklist and do it in bits and pieces.

**Comment by dankelley** [comment]
Checklist (which will be edited in place).

    *      oceApprox,
           oceColorsGebco,
           oceColorsPalette,
           oceColorsJet,
           oceColorsTwo,
           oceColors9A,
           oceColors9B,
    *      oceDebug,
    *      oceEdit,
    *      oceFilter,
           oceMagic,
    *      oceConvolve,
    *      oceContour,
    *      ocePmatch,
    *      oceSmooth,
    *      oceSpectrum,
    
    $ grep '^ *oce\.[a-z].*' NAMESPACE 
           oce.as.raw,
           oce.as.POSIXlt,
           oce.axis.POSIXct,
           oce.grid,
           oce.plot.ts,
           oce.write.table,

**Comment by richardsc** [comment]
Ok, maybe I'm a little confused again ... (it's not too hard). I thought we were proposing to use the "dot" names *only* for functions for which a similarly named function already exists (e.g. `oce.plot.ts()`), but that we'd stick with camelCase for all the other functions e.g. `oceColorsJet()`.

Re-reading your review list it now looks to me that you're proposing changing **ALL** `oce*()` functions from camelCase to dot notation, giving something like:

* ``oceApprox`` -> ``oce.approx``
* ``oceColorsGebco`` -> ``oce.colors.gebco``
* ``oceColorsPalette`` -> ``oce.colors.palette``
* ``oceColorsJet`` -> ``oce.colors.jet``
* ``oceColorsTwo`` -> ``oce.colors.two``
* ``oceColors9A`` -> ``oce.colors.9A``?? *I'm not sure what to do with this, and the next*
* ``oceColors9B`` -> ``oce.colors.9B``??
* ``oceDebug`` -> ``oce.debug``
* ``oceEdit`` -> ``oce.edit``
* ``oceFilter`` -> ``oce.filter``
* ``oceMagic`` -> ``oce.magic``
* ``oceConvolve`` -> ``oce.convolve``
* ``oceContour`` -> ``oce.contour``
* ``ocePmatch`` -> ``oce.pmatch``
* ``oceSmooth`` -> ``oce.smooth``
* ``oceSpectrum`` -> ``oce.spectrum``

Are these what you were thinking? I'm not sure that I like that functions like `oceColorsJet()` become `oce.colors.jet()`, but it might be silly to mix styles (e.g. `oce.colorsJet()`)

I guess it makes sense, for consistency. Actually, I don't *hate* the `oce.colorsJet()` form, which also works sensibly for `oce.colors9B()`.

P. S. I tried faceTime'ing you to chat quickly in case you were at your computer/phone, but no answer.

**Comment by dankelley** [comment]
Thanks, we're getting somewhere.

I feel that the first step is for me to write text for the vignette. That way, if the argument is unconvincing, we'll know that it is wrong. So, here is some possible text for the functions that start with ``oce.`` ...

> Functions in this package generally use a camel-case notation [citation], i.e. if a function chops wood and then piles it, the name might be ``chopPile``. An exception is in function names that would otherwise conflict with an existing function in either the base R system or in popular packages. Thus, for example, ``oceApprox`` is so named to avoid a conflict with the (somewhat similar) base function ``approx``. This is a matter of convenience for the user, who otherwise would have to remember to write ``oce::approx`` to get the function provided by oce.

If you would like to suggest alternatives, can you put them within this comment, *in place*, please? (I get confused when I have to scroll a browser to compare things.)

OK, back to the list. I've put my ideas below, with some notes. I wonder if we can make progress by having you edit (in place) putting any arguments in the same item, maybe prefixed by "CR" so we can see it easily?

My list contains two notes. You could add more here (in place).

**NOTE 1** We have one outlier in the list, namely the  ``oceColors*`` functions. When I started, I probably had an idea for ``oceColors``, without the extra nuances. And I still think that would be good, e.g. we could say ``oce.colors(..., scheme="jet")`` or something. (I should code this.) So now we have a reason for the dotted form. Well, what about the rest: to dot or not to dot? I feel the rest should be camel case because that's what oce does for functions. Thus I argue for ``oce.colorsGebco``, and that's what I've put in my list below.

**NOTE 2** The point is not that we are providing exactly an alternative function (which is all shades of grey, anyhow) but that the other name is already taken. Thus, ``oce.debug`` or ``oceDebug`` cannot be written ``debug`` because doing so would conflict.

* ``oceApprox`` -> ``oce.approx``
* ``(nothing)`` -> ``oce.colors`` a new function as described in NOTE 1
* ``oceColorsGebco`` -> ``oce.colorsGebco`` NOTE 1
* ``oceColorsPalette`` -> ``oce.colorsPalette`` NOTE 1
* ``oceColorsJet`` -> ``oce.colorsJet`` NOTE 1 and 3
* ``oceColorsTwo`` -> ``oce.colorsTwo`` NOTE 1
* ``oceColors9A`` -> ``oce.colors9A`` NOTE 1
* ``oceColors9B`` -> ``oce.colors9B`` NOTE 1
* ``oceDebug`` -> ``oce.debug``
* ``oceEdit`` -> ``oce.edit``  NOTE 2
* ``oceFilter`` -> ``oce.filter``
* ``oceMagic`` -> ``oce.magic``
* ``oceConvolve`` -> ``oce.convolve``
* ``oceContour`` -> ``oce.contour``
* ``ocePmatch`` -> ``oce.pmatch``
* ``oceSmooth`` -> ``oce.smooth``
* ``oceSpectrum`` -> ``oce.spectrum``

**Comment by dankelley** [comment]
I have added 1/4 page to the vignette on function names. I'm snapshotting below (which saves a build). Please feel free to comment here, to alter the oce.Rnw file, or to use goodreader for hand-written notes (this last being my preferred scheme).

![screen shot 2015-02-08 at 10 38 42 am](https://cloud.githubusercontent.com/assets/99469/6096925/f2d9a786-af7e-11e4-9cef-14b8fe059623.png)

**Comment by dankelley** [comment]
I've done ``oceApprox`` as a test case. It is quite simple. The recoding took about 5 minutes, while the test build took 10 minutes (on a fast machine). Coding this sort of thing is greatly aided by grep, e.g. the way I found the files to edit was
```bash
mvim $(git grep -l oceApprox)
```

**Comment by richardsc** [comment]
This all looks good to me. 

I like the idea of an `oce.colors()` function (as long as we still keep the other aliases).

And I agree that for consistency `oceMagic` should be aliased to `oce.magic()`.

**Comment by dankelley** [comment]
I think I've cleaned this all up in ``develop``, so it should be ready for some testing. I haven't written a new ``oce.colors()`` function yet, but that could come later.

**Comment by richardsc** [comment]
Great!

What do you propose is the best way to test? Perhaps come up with an example for each, and then run that example using both forms of the functions?

**Comment by dankelley** [comment]
I mean more "use" than "test", I think. I changed all occurrences in my book, and the book still builds OK. Also, ``oce`` builds and checks OK, which means that the functions are found in the namespace, that none of the help pages are referring to a non-existent function, and so forth.

Maybe the best plan would be just to close the issue, since it (as defined by the title) has been addressed.  New issues could be opened for flaws in the implementation.

I really like the "look" of these function names better.
